pimd, pim6d: changes to pimreg register socket initialization

routingrocks opened the pull request:
We separated PIM register messages to send out only on the reg_sock socket. This is to handle a case where the FHR is RP: we send out a register message and receive a reg stop in a pim_sock_fd. This is fine, but the problem arises when we are in a multi-VRF setup and FHR is RP. When an pimreg interface moves from one VRF to another, the socket is lost. and, when we create a PIM register interface, this interface is never bound to the non default vrf's reg_sock socket.

With this fix, when FHR is a RP, the non-default VRF reg_sock will now bind to the interface. As a result, the PIM register message will be sent out.

Testing Done:
Verified mroute on RP/FHR and S,G programmed, traffic reaches the receiver

Signed-off-by: Rajesh Varatharaj <rvaratharaj@nvidia.com>

routingrocks committed:
pimd, pim6d: changes to pimreg register socket initialization

We separated PIM register messages to send out only on the reg_sock socket.
This is to handle a case where the FHR is RP: we send out a register message
and receive a reg stop in a pim_sock_fd. This is fine, but the problem arises
when we are in a multi-VRF setup and FHR is RP. When an pimreg interface moves
from one VRF to another, the socket is lost. and, when we create a PIM register
interface, this interface is never bound to the non default vrf's reg_sock socket.

With this fix, when FHR is a RP, the non-default VRF reg_sock will
now bind to the PIM register interface. As a result,
the PIM register message will be sent out.

Testing Done:
Verified mroute on RP/FHR and S,G programmed, traffic reaches the receiver

Signed-off-by: Rajesh Varatharaj <rvaratharaj@nvidia.com>

routingrocks commented:
2023/04/03 17:58:54.306924 PIM: [QS4SS-V6MR6] pim_mroute_msg: pim kernel upcall WRONGVIF type=2 ip_p=0 from fd=16 for (S,G)=(20.20.10.1,229.1.2.0) on swp1s2 vifi=5  size=28 reg_sock fd=54
2023/04/03 17:58:54.306926 PIM: [M6BDE-64MCQ] pim_mroute_msg_wrongvif: WRONGVIF (S,G)=(20.20.10.1,229.1.2.0) could not find channel on interface swp1s2
2023/04/03 17:58:54.306929 PIM: [KXPG5-HXYPC] pim_mroute_msg_wrongvif: WRONGVIF (*,G)=(*,229.1.2.0) could not find channel on interface swp1s2
2023/04/03 17:58:54.306932 PIM: [J9844-9Y88F]  vif=5
2023/04/03 17:58:54.306933 PIM: [QS4SS-V6MR6] pim_mroute_msg: pim kernel upcall WRVIFWHOLE type=4 ip_p=0 from fd=16 for (S,G)=(20.20.10.1,229.1.2.0) on swp1s2 vifi=5  size=56 reg_sock fd=54
2023/04/03 17:58:54.306941 PIM: [XXYFJ-4M8ZW] Lookedup(229.1.2.0/32): rn 0x556fe5e44d70 found:229.1.2.0/24
2023/04/03 17:58:54.306943 PIM: [HWBEV-9D9K2] pim_rpf_set_refresh_time: vrf(red) New last route change time: 1680544734306942
2023/04/03 17:58:54.306947 PIM: [XXYFJ-4M8ZW] Lookedup(229.1.2.0/32): rn 0x556fe5e44d70 found:229.1.2.0/24
2023/04/03 17:58:54.306953 PIM: [W5DFK-HMAF0] pim_mroute_msg_wholepkt:  Sending Register Packet  on swp1s2 reg_sock fd=54
2023/04/03 17:58:54.306956 PIM: [XXYFJ-4M8ZW] Lookedup(229.1.2.0/32): rn 0x556fe5e44d70 found:229.1.2.0/24
2023/04/03 17:58:54.306957 PIM: [HWBEV-9D9K2] pim_rpf_set_refresh_time: vrf(red) New last route change time: 1680544734306957
2023/04/03 17:58:54.306958 PIM: [P0N0W-ZMZYH] pim_mroute_msg_wholepkt:  Sending (20.20.10.1,229.1.2.0) Register Packet 2 to 70.0.0.7 on swp1s2 reg_sock fd=54
2023/04/03 17:58:54.306961 PIM: [PCAG1-GREKK] **Sending (20.20.10.1,229.1.2.0) Register Packet to 70.0.0.7**
2023/04/03 17:58:54.306963 PIM: [MXS69-VR5QP] pim_register_send:  Sending (20.20.10.1,229.1.2.0) Register Packet to 70.0.0.7 on red reg sock fd 54 ; pim sock fd = 21
2023/04/03 17:58:54.306966 PIM: [GVEMA-2S3KA] pim_msg_send: to 70.0.0.7 on red: msg_size=44 checksum=ffde fd=54
2023/04/03 17:58:54.306968 PIM: [Q7SDG-9TB27] pim_msg_send: pkt dump size=44
2023/04/03 17:58:54.306972 PIM: [Q79JG-SJBP4] 00007ffc43d41c90: 21 00 de ff 00 00 00 00 !.......
2023/04/03 17:58:54.306974 PIM: [Q79JG-SJBP4] 00007ffc43d41c98: 45 00 00 24 c2 6f 40 00 E..$.o@.
2023/04/03 17:58:54.306975 PIM: [Q79JG-SJBP4] 00007ffc43d41ca0: 14 11 9f 43 14 14 0a 01 ...C....
2023/04/03 17:58:54.306976 PIM: [Q79JG-SJBP4] 00007ffc43d41ca8: e5 01 02 00 e3 66 03 e8 .....f..
2023/04/03 17:58:54.306977 PIM: [Q79JG-SJBP4] 00007ffc43d41cb0: 00 10 bb f4 66 6f 6f 62 ....foob
2023/04/03 17:58:54.306979 PIM: [Q79JG-SJBP4] 00007ffc43d41cb8: 61 72 20 30             ar 0
2023/04/03 17:58:54.306987 PIM: [RRBDT-6NWPY] **pim_msg_send_frame: sendto() failure to 70.0.0.7: iface=red fd=54 msg_size=64: Network is unreachable**

routingrocks commented:
ci:rerun

ton31337 commented on pimd/pim_iface.c:
```diff
@@ -1536,6 +1536,12 @@ void pim_if_create_pimreg(struct pim_instance *pim)
 				 pim->vrf->data.l.table_id);
 
+		if (pim_reg_sock_bind(pim)) {
+			zlog_warn(
```
Do we print zlog_warn() twice if pim_socket_bind() fails?

routingrocks replied:
Agree, I will remove the one in the caller

routingrocks committed:
pimd: capture pim_msg_send_frame return

It's valuable to capture the return from `pim_msg_send()`
 and log the error. Otherwise, send failures will be silently ignored.

Signed-off-by: Rajesh Varatharaj <rvaratharaj@nvidia.com>

donaldsharp commented on pimd/pim_pim.c:
```diff
@@ -717,8 +718,10 @@ int pim_msg_send(int fd, pim_addr src, pim_addr dst, uint8_t *pim_msg,
-			   tolen, ifp->name);
+	ret = pim_msg_send_frame(fd, (char *)buffer, sendlen,
+				 (struct sockaddr *)&to, tolen, ifp->name);
+	if (ret)
```
why not just return ret?  in both functions?

mobash-rasool commented on pimd/pim_sock.c:
```diff
@@ -418,3 +418,25 @@ int pim_socket_getsockname(int fd, struct sockaddr *name, socklen_t *namelen)
+int pim_reg_sock_bind(struct pim_instance *pim)
+{
+	if (!pim)
+		return 0;
```
Can we use PIM_SOCK_ERR_NONE to make errors consistent in this function?

mobash-rasool commented on pimd/pim_pim.c:
```diff
@@ -717,19 +718,19 @@ int pim_msg_send(int fd, pim_addr src, pim_addr dst, uint8_t *pim_msg,
-	pim_msg_send_frame(fd, (char *)buffer, sendlen, (struct sockaddr *)&to,
-			   tolen, ifp->name);
-	return 0;
+	ret = pim_msg_send_frame(fd, (char *)buffer, sendlen,
```
Do we need a ret variable here? We can directly return the function. Otherwise coverity or some static tool will complain about not using ret variable.
return pim_msg_send_frame(fd, (char *)buffer, sendlen, (struct sockaddr *)&to, tolen, ifp->name);

mobash-rasool commented on pimd/pim_pim.c:
```diff
@@ -717,19 +718,19 @@ int pim_msg_send(int fd, pim_addr src, pim_addr dst, uint8_t *pim_msg,
 	iovector[0].iov_len = pim_msg_size;
 
-	pim_msg_send_frame(src, dst, ifp->ifindex, &iovector[0], fd);
+	ret = pim_msg_send_frame(src, dst, ifp->ifindex, &iovector[0], fd);
```
same as above.

mobash-rasool commented on pimd/pim_pim.c:
```diff
@@ -637,6 +637,7 @@ int pim_msg_send(int fd, pim_addr src, pim_addr dst, uint8_t *pim_msg,
 		 int pim_msg_size, struct interface *ifp)
 {
 	struct pim_interface *pim_ifp;
+	int ret;
```
this should be deleted.

mobash-rasool commented on pimd/pim_iface.c:
```diff
@@ -1536,6 +1536,7 @@ void pim_if_create_pimreg(struct pim_instance *pim)
 			snprintf(pimreg_name, sizeof(pimreg_name), PIMREG "%u",
 				 pim->vrf->data.l.table_id);
 
+		pim_reg_sock_bind(pim);
```
This function pim_reg_sock_bind returns int but it is not used here.

routingrocks commented:
ci:rerun

mobash-rasool reviewed (changes requested):
Please add a topotest to make sure this works now and in future too.

routingrocks commented:
@mobash-rasool here is the topotest https://github.com/FRRouting/frr/pull/13619

mobash-rasool commented:
@routingrocks:  Shouldn't the test case in #13619 fail since that PR does not contain your fix?

ton31337 commented:
@frrbot rereview

ton31337 commented on pimd/pim_sock.c:
```diff
@@ -418,3 +418,25 @@ int pim_socket_getsockname(int fd, struct sockaddr *name, socklen_t *namelen)
+			return;
+		}
+	}
+	return;
```
This is useless, implicit return by void.

routingrocks commented:
confict addressed in `pimd/pim_pim.c`

riw777 commented:
@frrbot autoclose in 2 weeks
